src/nameservice/ClickNameService.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+pragma solidity ^0.8.23;
+
+import {ERC721} from "openzeppelin-contracts/contracts/token/ERC721/ERC721.sol";
+import {ERC721Burnable} from "openzeppelin-contracts/contracts/token/ERC721/extensions/ERC721Burnable.sol";
+import {AccessControl} from "openzeppelin-contracts/contracts/access/AccessControl.sol";
+import {Strings} from "openzeppelin-contracts/contracts/utils/Strings.sol";
+import {IClickNameService} from "./IClickNameService.sol";
+
+/**
+ * @title ClickNameService based on ClaveNameService which is authored by https://getclave.io
+ * @notice L2 name service contract that is built compatible to resolved as ENS subdomains by L2 resolver
+ * @dev Names can only be registered by authorized accounts
+ * @dev Addresses can only have one name at a time
+ * @dev Subdomains are stored as ERC-721 assets, cannot be transferred
+ * @dev If renewals are enabled, non-renewed names can be burnt after expiration timeline
+ */
+contract ClickNameService is IClickNameService, ERC721Burnable, AccessControl {
+    using Strings for uint256;
+
+    struct NameOwner {
+        address owner;
+        string name;
+    }
+
+    // Role to be authorized as default minter
+    bytes32 public constant REGISTERER_ROLE = keccak256("REGISTERER_ROLE");
+    // Default expiration duration
+    uint256 public expiryDuration = 365 days;
+
+    // token id to expires timestamp
+    mapping(uint256 => uint256) public expires;
+
+    event NameRegistered(string indexed name, address indexed owner, uint256 expires);
+
+    constructor(address admin, address registrar) ERC721("ClickNameService", "CLK") {
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+        _grantRole(REGISTERER_ROLE, registrar);
+    }
+
+    /// @inheritdoc IClickNameService
+    function resolve(string memory name) external view returns (address) {
+        uint256 tokenId = uint256(keccak256(abi.encodePacked(name)));
+        address owner = ownerOf(tokenId);
+        require(expires[tokenId] > block.timestamp, "Name expired.");
+        return owner;
+    }
+
+    /**
+     * @notice Register multiple names at once
+     */
+    function batchRegister(NameOwner[] memory nameOwners) external isAuthorized {
```

Comment:
Not a big fan of modifiers, we should go back to our normal pattern (this also applies to pretty much most functions below)

Reply:
done